docs/boot_modes.md

```diff
@@ -0,0 +1,111 @@
+# GardenLinux Boot Modes :potted_plant::penguin:
+
+GardenLinux supports two primary boot modes: Legacy and Unified System Image (USI).
+
+Legacy is the traditional way most Linux distros boot and also how all GardenLinux versions prior to 17?? used to boot.
+Legacy boot images are dual mode images, working both with UEFI boot and BIOS boot.
+
+USI mode is a novel approach, where the entire system gets embedded into a single UKI binary.
+This brings several advantages, such as allowing for in-place updates and enhanced security, but is only compatible with UEFI boot.
+
+|                          | USI                | Legacy             |
+|--------------------------|--------------------|--------------------|
+| UEFI boot                | :white_check_mark: | :white_check_mark: |
+| BIOS boot                | :x:                | :white_check_mark: |
+| secureboot / trustedboot | :heavy_check_mark: | :x:                |
+| in-place updates         | :white_check_mark: | :x:                |
+| TPM 2.0 disk encryption  | :heavy_check_mark: | :x:                |
+
+*( :heavy_check_mark: = optional )*
+
+> [!TIP]
+> If you are unsure which mode to pick: go for the *trustedboot* images, these are USIs with all optional security features enabled.
+
+## Legacy
+
+Legacy boot mode uses a standard partition layout.
+There is a boot partition, used both by syslinux for BIOS boot and systemd-boot for UEFI boot.
+The rootfs is on a simple ext4 partition, it is fully mutable.
+
+Legacy mode does not attempt to provide any protection against offline attacks.
+
+```shell
+# example disk layout for legacy mode
+
+vda                                                                         
+├─vda1
+│    vfat   FAT32 EFI   C1FE-FA19                                28M    67% /efi
+└─vda2
+     ext4   1.0   ROOT  aea6d56e-a2bf-944c-78fe-184a4d6add9f   29.1G     2% /
+```
+
+## Unified System Image (USI)
+
+The design is loosely inspired by Lennart Poettering's ideas[^1] on how modern operating systems should boot.
+We also fully embrace Unified Kernel Images (UKIs)[^2] and we also make use of an immutable root filesystem.
+However, the design differs in how the root filesystem is verified.
+
+In Lennart Poettering's design, the root (or just the `/usr` partition) is stored on a dm-verity disk; the root hash of the dm-verity Merkle tree is then embedded into the UKI.
+While this design works well, it does add quite a bit of complexity.
+It requires at least three partitions (more if you want in-place updates with A/B partitions): EFI system partition, root dm-verity data partition, and root dm-verity hash tree partition.
+
+Due to the incredibly small footprint of our root partition, we can afford to go a different route: we can embed the entire root filesystem into the UKI.
```

Comment:
> incredibly small footprint

maybe give some numbers here. the limit is the 4gi limit of the FAT on EFI, correct?

Reply:
correct, we should be somewhere around 400-500MiB for the uncompressed rootfs which drops down to just over 200MiB for the lz4 compressed EROFS, resulting in a final UKI of under 200MiB (due to secondary compression with xz). So far away from the 4GiB limit.

Eventually the EROFS will/should be zstd compressed directly, but support for this just landed in the kernel recently(https://github.com/torvalds/linux/commit/7c35de4df1056a5a1fb4de042197b8f5b1033b61)